{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal\n",
    "\n",
    "* Compare the metadata diversity in Cell x Gene versus scBaseCamp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_uri = \"/scratch/multiomics/nickyoungblut/tiledb-loader/tiledb-soma_GeneFull_Ex50pAS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotnine as pn\n",
    "import tiledbsoma\n",
    "import tiledbsoma.io\n",
    "import cellxgene_census"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX10681588</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>not specified</td>\n",
       "      <td>7576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX18700948</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>severe dengue</td>\n",
       "      <td>6823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX18700950</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>severe dengue</td>\n",
       "      <td>4385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX18700949</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>severe dengue</td>\n",
       "      <td>1928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX19713411</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>human</td>\n",
       "      <td>adipose differentiating cells</td>\n",
       "      <td>none</td>\n",
       "      <td>5879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21189</th>\n",
       "      <td>SRX10664197</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21190</th>\n",
       "      <td>SRX5317413</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21191</th>\n",
       "      <td>SRX11523475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21192</th>\n",
       "      <td>SRX10188985</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21193</th>\n",
       "      <td>SRX5545323</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21194 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SRX_accession      lib_prep organism                         tissue  \\\n",
       "0       SRX10681588  10x_Genomics    human                          blood   \n",
       "1       SRX18700948  10x_Genomics    human                          blood   \n",
       "2       SRX18700950  10x_Genomics    human                          blood   \n",
       "3       SRX18700949  10x_Genomics    human                          blood   \n",
       "4       SRX19713411  10x_Genomics    human  adipose differentiating cells   \n",
       "...             ...           ...      ...                            ...   \n",
       "21189   SRX10664197  10x_Genomics      NaN                            NaN   \n",
       "21190    SRX5317413  10x_Genomics      NaN                            NaN   \n",
       "21191   SRX11523475           NaN      NaN                            NaN   \n",
       "21192   SRX10188985  10x_Genomics      NaN                            NaN   \n",
       "21193    SRX5545323  10x_Genomics      NaN                            NaN   \n",
       "\n",
       "             disease  count_all  \n",
       "0      not specified       7576  \n",
       "1      severe dengue       6823  \n",
       "2      severe dengue       4385  \n",
       "3      severe dengue       1928  \n",
       "4               none       5879  \n",
       "...              ...        ...  \n",
       "21189            NaN      12171  \n",
       "21190            NaN       5206  \n",
       "21191            NaN       6567  \n",
       "21192            NaN       9704  \n",
       "21193            NaN       5316  \n",
       "\n",
       "[21194 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load scBaseCamp data\n",
    "target_cols = [\"SRX_accession\", \"lib_prep\", \"organism\", \"tissue\", \"disease\"]\n",
    "\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    scbc_metadata = (\n",
    "        exp.obs.read(column_names=target_cols)\n",
    "        .concat()\n",
    "        .group_by(target_cols)\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .to_pandas()\n",
    "    )\n",
    "scbc_metadata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_general</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>disease</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>count_all</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0d2ee4ac-05ee-40b2-afb6-ebb584caa867</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung</td>\n",
       "      <td>cell</td>\n",
       "      <td>normal</td>\n",
       "      <td>classical monocyte</td>\n",
       "      <td>198</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0d2ee4ac-05ee-40b2-afb6-ebb584caa867</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung</td>\n",
       "      <td>cell</td>\n",
       "      <td>normal</td>\n",
       "      <td>respiratory goblet cell</td>\n",
       "      <td>62</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0d2ee4ac-05ee-40b2-afb6-ebb584caa867</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung</td>\n",
       "      <td>cell</td>\n",
       "      <td>normal</td>\n",
       "      <td>basal cell</td>\n",
       "      <td>181</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0d2ee4ac-05ee-40b2-afb6-ebb584caa867</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung</td>\n",
       "      <td>cell</td>\n",
       "      <td>normal</td>\n",
       "      <td>pulmonary alveolar type 2 cell</td>\n",
       "      <td>749</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0d2ee4ac-05ee-40b2-afb6-ebb584caa867</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>lung</td>\n",
       "      <td>lung</td>\n",
       "      <td>cell</td>\n",
       "      <td>normal</td>\n",
       "      <td>capillary endothelial cell</td>\n",
       "      <td>339</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51176</th>\n",
       "      <td>4807</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>sci-RNA-seq3</td>\n",
       "      <td>embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>normal</td>\n",
       "      <td>club cell</td>\n",
       "      <td>6264</td>\n",
       "      <td>mus_musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51177</th>\n",
       "      <td>4808</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>sci-RNA-seq3</td>\n",
       "      <td>embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>normal</td>\n",
       "      <td>activated CD4-negative, CD8-negative type I NK...</td>\n",
       "      <td>6621</td>\n",
       "      <td>mus_musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51178</th>\n",
       "      <td>4809</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>sci-RNA-seq3</td>\n",
       "      <td>embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>normal</td>\n",
       "      <td>pulmonary alveolar type 1 cell</td>\n",
       "      <td>14321</td>\n",
       "      <td>mus_musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51179</th>\n",
       "      <td>4810</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>sci-RNA-seq3</td>\n",
       "      <td>embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>normal</td>\n",
       "      <td>pulmonary alveolar type 2 cell</td>\n",
       "      <td>21443</td>\n",
       "      <td>mus_musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51180</th>\n",
       "      <td>4811</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>sci-RNA-seq3</td>\n",
       "      <td>embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>normal</td>\n",
       "      <td>retinal rod cell</td>\n",
       "      <td>2743</td>\n",
       "      <td>mus_musculus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51181 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index                            dataset_id         assay  tissue  \\\n",
       "0          0  0d2ee4ac-05ee-40b2-afb6-ebb584caa867    Smart-seq2    lung   \n",
       "1          1  0d2ee4ac-05ee-40b2-afb6-ebb584caa867    Smart-seq2    lung   \n",
       "2          2  0d2ee4ac-05ee-40b2-afb6-ebb584caa867    Smart-seq2    lung   \n",
       "3          3  0d2ee4ac-05ee-40b2-afb6-ebb584caa867    Smart-seq2    lung   \n",
       "4          4  0d2ee4ac-05ee-40b2-afb6-ebb584caa867    Smart-seq2    lung   \n",
       "...      ...                                   ...           ...     ...   \n",
       "51176   4807  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3  sci-RNA-seq3  embryo   \n",
       "51177   4808  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3  sci-RNA-seq3  embryo   \n",
       "51178   4809  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3  sci-RNA-seq3  embryo   \n",
       "51179   4810  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3  sci-RNA-seq3  embryo   \n",
       "51180   4811  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3  sci-RNA-seq3  embryo   \n",
       "\n",
       "      tissue_general suspension_type disease  \\\n",
       "0               lung            cell  normal   \n",
       "1               lung            cell  normal   \n",
       "2               lung            cell  normal   \n",
       "3               lung            cell  normal   \n",
       "4               lung            cell  normal   \n",
       "...              ...             ...     ...   \n",
       "51176         embryo         nucleus  normal   \n",
       "51177         embryo         nucleus  normal   \n",
       "51178         embryo         nucleus  normal   \n",
       "51179         embryo         nucleus  normal   \n",
       "51180         embryo         nucleus  normal   \n",
       "\n",
       "                                               cell_type  count_all  \\\n",
       "0                                     classical monocyte        198   \n",
       "1                                respiratory goblet cell         62   \n",
       "2                                             basal cell        181   \n",
       "3                         pulmonary alveolar type 2 cell        749   \n",
       "4                             capillary endothelial cell        339   \n",
       "...                                                  ...        ...   \n",
       "51176                                          club cell       6264   \n",
       "51177  activated CD4-negative, CD8-negative type I NK...       6621   \n",
       "51178                     pulmonary alveolar type 1 cell      14321   \n",
       "51179                     pulmonary alveolar type 2 cell      21443   \n",
       "51180                                   retinal rod cell       2743   \n",
       "\n",
       "           organism  \n",
       "0      homo_sapiens  \n",
       "1      homo_sapiens  \n",
       "2      homo_sapiens  \n",
       "3      homo_sapiens  \n",
       "4      homo_sapiens  \n",
       "...             ...  \n",
       "51176  mus_musculus  \n",
       "51177  mus_musculus  \n",
       "51178  mus_musculus  \n",
       "51179  mus_musculus  \n",
       "51180  mus_musculus  \n",
       "\n",
       "[51181 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load cell x gene obs metadata\n",
    "target_cols = [\"dataset_id\", \"assay\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\", \"cell_type\"]\n",
    "\n",
    "cxg_metadata = []\n",
    "for organism in [\"homo_sapiens\", \"mus_musculus\"]:\n",
    "    with cellxgene_census.open_soma(census_version=\"2025-01-30\") as census:\n",
    "        df = (\n",
    "            census[\"census_data\"][organism].obs\n",
    "            .read(column_names = target_cols).concat()\n",
    "            .group_by(target_cols).aggregate([([], 'count_all')]).to_pandas()\n",
    "        )\n",
    "        df[\"organism\"] = organism\n",
    "        cxg_metadata.append(df)\n",
    "\n",
    "cxg_metadata = pd.concat(cxg_metadata).reset_index()\n",
    "cxg_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import cellxgene_census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with cellxgene_census.open_soma() as census:\n",
    "    print(census[\"census_data\"][\"homo_sapiens\"].obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with cellxgene_census.open_soma() as census:\n",
    "    print(census[\"census_data\"][\"mus_musculus\"].obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_cols = [\"dataset_id\", \"assay\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]\n",
    "\n",
    "with cellxgene_census.open_soma(census_version=\"2025-01-30\") as census:\n",
    "    cxg_metadata = pd.concat([\n",
    "        census[\"census_data\"][\"homo_sapiens\"].obs.read(column_names = target_cols).concat().to_pandas(),\n",
    "        census[\"census_data\"][\"mus_musculus\"].obs.read(column_names = target_cols).concat().to_pandas(),\n",
    "    ])\n",
    "\n",
    "cxg_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tiledb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
